apis/r/src/metadata.cpp

```diff
@@ -0,0 +1,179 @@
+//' @export
+// [[Rcpp::export]]
+int32_t get_metadata_num(std::string& uri, bool is_array, Rcpp::XPtr<somactx_wrap_t> ctxxp) {
+    // shared pointer to SOMAContext from external pointer wrapper
+    std::shared_ptr<tdbs::SOMAContext> sctx = ctxxp->ctxptr;
+    // SOMA Object unique pointer (aka soup)
+    auto soup = getObjectUniquePointer(is_array, OpenMode::read, uri, sctx);
+    int32_t nb = soup->metadata_num();
+    return nb;
+}
+
+//' Read all metadata (as named list)
+//'
+//' This function currently supports metadata as either a string or an 'int64' (or 'int32').
+//' It will error if a different datatype is encountered.
+//' @param uri The array URI
+//' @param ctxxp An external pointer to the SOMAContext wrapper
+//' @export
+// [[Rcpp::export]]
+Rcpp::List get_all_metadata(std::string& uri, bool is_array, Rcpp::XPtr<somactx_wrap_t> ctxxp) {
+    // shared pointer to SOMAContext from external pointer wrapper
+    std::shared_ptr<tdbs::SOMAContext> sctx = ctxxp->ctxptr;
+
+    // SOMA Object unique pointer (aka soup)
+    auto soup = getObjectUniquePointer(is_array, OpenMode::read, uri, sctx);
+    auto mvmap = soup->get_metadata();
+
+    std::vector<std::string> namvec;
+    Rcpp::List lst;
+    for (auto it = mvmap.begin(); it != mvmap.end(); it++) {
+        std::string key = it->first;
+        namvec.push_back(key);
+        tdbs::MetadataValue val = it->second;
+        auto dtype = std::get<0>(val);
+        auto len = std::get<1>(val);
+        const void* ptr = std::get<2>(val);
+        if (dtype == TILEDB_STRING_UTF8 || dtype == TILEDB_STRING_ASCII) {
+            auto str = std::string((char*) ptr, len);
+            lst.push_back(str);
+        } else if (dtype == TILEDB_INT64) {
+            std::vector<int64_t> v(len);
+            std::memcpy(&(v[0]), ptr, len*sizeof(int64_t));
+            lst.push_back(Rcpp::toInteger64(v));
+        } else if (dtype == TILEDB_INT32) {
+            Rcpp::IntegerVector v(len);
+            std::memcpy(v.begin(), ptr, len*sizeof(int32_t));
+        } else {
+            auto txt = tiledb::impl::type_to_str(dtype);
+            Rcpp::stop("Currently unsupported type '%s'", txt.c_str());
+        }
+    }
+    lst.attr("names") = Rcpp::CharacterVector(namvec.begin(), namvec.end());
+    return lst;
+}
+
+//' Read metadata (as a string)
+//'
+//' @param uri The array URI
+//' @param key The array metadata key
+//' @param ctxxp An external pointer to the SOMAContext wrapper
+//' @export
+// [[Rcpp::export]]
+std::string get_metadata(std::string& uri, std::string& key, bool is_array,
+                         Rcpp::XPtr<somactx_wrap_t> ctxxp) {
+    // shared pointer to SOMAContext from external pointer wrapper
+    std::shared_ptr<tdbs::SOMAContext> sctx = ctxxp->ctxptr;
+
+    // SOMA Object unique pointer (aka soup)
+    auto soup = getObjectUniquePointer(is_array, OpenMode::read, uri, sctx);
+    auto mv = soup->get_metadata(key);
```

Comment:
Nevermind I think I understand now. You open the `SOMAObject` and then call `SOMAObject::get_metadata` which looks how it was also implemented before when you had it calling `tiledb::tiledb_put_metadata` calls. You are just switching the internals from using TileDB to SOMA C++ but it still looks the same to the user.

Reply:
> Am I overlooking something?

No I think I just misunderstood how to read the R code

Reply:
Phew.  All good then.

And yes given the 11800 lines of R code (!!) in apis/r/R/*, never mind the 10300 lines of test code, I am trying to stick with existing interfaces and "swap" out calling tiledb core with calling libtiledbsoma.